File: go/internal/worker/worker.go
package worker

import (
	"time"

	"payment-processing/internal/fraud"
	"payment-processing/internal/processor"
	"payment-processing/internal/retry"
	"payment-processing/internal/store"
	"payment-processing/internal/webhooks"
)

type Worker struct {
	Store          *store.InMemoryStore
	WebhookSecret  string
	MaxAttempts    int
	BaseBackoff    time.Duration
	PollInterval   time.Duration
}

func (w *Worker) Run() {
	for {
		task, ok := w.Store.DequeueTask()
		if !ok {
			time.Sleep(w.PollInterval)
			continue
		}
		switch task.Type {
		case "process_payment":
			w.handleProcessPayment(task.IntentID)
		case "deliver_webhooks":
			w.handleDeliverWebhooks(task.EventID)
		}
	}
}

func (w *Worker) handleProcessPayment(intentID string) {
	intent, ok := w.Store.GetPaymentIntent(intentID)
	if !ok || intent.Status != "processing" {
		return
	}
	score, decision, signals := fraud.Evaluate(intent)
	w.Store.CreateFraudAssessment(intent.ID, score, decision, signals)
	if decision == "block" {
		w.Store.UpdatePaymentIntent(intent.ID, map[string]interface{}{"status": "failed"})
		w.Store.CreateEvent(intent.MerchantID, "payment_intent.failed", map[string]interface{}{
			"id":     intent.ID,
			"status": "failed",
			"reason": "fraud_block",
		})
		return
	}
	client := processor.Client{}
	processorRef, err := retry.WithRetries(
		func() (string, error) {
			return client.Authorize(intent.Amount, intent.Currency, intent.PaymentMethodToken)
		},
		w.MaxAttempts,
		w.BaseBackoff,
		func(err error) bool { return err == processor.ErrRetryable },
	)
	if err != nil {
		reason := "processor_unavailable"
		if err == processor.ErrDeclined {
			reason = "declined"
		}
		w.Store.UpdatePaymentIntent(intent.ID, map[string]interface{}{"status": "failed"})
		w.Store.CreateEvent(intent.MerchantID, "payment_intent.failed", map[string]interface{}{
			"id":     intent.ID,
			"status": "failed",
			"reason": reason,
		})
		return
	}
	charge := w.Store.CreateCharge(intent.ID, intent.Amount, intent.Currency, "succeeded")
	w.Store.UpdatePaymentIntent(intent.ID, map[string]interface{}{
		"status":              "succeeded",
		"charge_id":           charge.ID,
		"processor_reference": processorRef,
	})
	w.Store.CreateEvent(intent.MerchantID, "payment_intent.succeeded", map[string]interface{}{
		"id":        intent.ID,
		"status":    "succeeded",
		"charge_id": charge.ID,
	})
}

func (w *Worker) handleDeliverWebhooks(eventID string) {
	event, ok := w.Store.Events[eventID]
	if !ok {
		return
	}
	endpoints := w.Store.ListWebhookEndpoints(event.MerchantID)
	payload := map[string]interface{}{
		"id":         event.ID,
		"type":       event.Type,
		"created_at": event.CreatedAt,
		"data":       event.Data,
	}
	for _, endpoint := range endpoints {
		_, err := retry.WithRetries(
			func() (string, error) {
				return "ok", webhooks.Deliver(endpoint.URL, w.WebhookSecret, payload)
			},
			w.MaxAttempts,
			w.BaseBackoff,
			func(err error) bool { return true },
		)
		if err != nil {
			w.Store.UpdateEventStatus(event.ID, "failed")
			return
		}
	}
	w.Store.UpdateEventStatus(event.ID, "delivered")
}

File: typescript/src/worker/worker.ts
import { config } from "../common/config.js";
import { evaluateFraud } from "../common/fraud.js";
import { ProcessorClient, ProcessorDeclinedError, RetryableProcessorError } from "../common/processor.js";
import { withRetries } from "../common/retry.js";
import { store } from "../common/store.js";
import { deliverWebhook } from "../common/webhooks.js";

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

const processPayment = async (intentId: string) => {
  const intent = store.getPaymentIntent(intentId);
  if (!intent || intent.status !== "processing") return;

  const { score, decision, signals } = evaluateFraud(intent);
  store.createFraudAssessment(intent.id, score, decision, signals);

  if (decision === "block") {
    store.updatePaymentIntent(intent.id, { status: "failed" });
    store.createEvent(intent.merchantId, "payment_intent.failed", {
      id: intent.id,
      status: "failed",
      reason: "fraud_block"
    });
    return;
  }

  const processor = new ProcessorClient();
  try {
    const reference = await withRetries(
      async () => processor.authorize(intent.amount, intent.paymentMethodToken),
      config.maxRetryAttempts,
      config.baseBackoffMs,
      (err) => err instanceof RetryableProcessorError
    );
    const charge = store.createCharge(intent.id, intent.amount, intent.currency, "succeeded");
    store.updatePaymentIntent(intent.id, {
      status: "succeeded",
      chargeId: charge.id,
      processorReference: reference
    });
    store.createEvent(intent.merchantId, "payment_intent.succeeded", {
      id: intent.id,
      status: "succeeded",
      charge_id: charge.id
    });
  } catch (error) {
    const reason = error instanceof ProcessorDeclinedError ? "declined" : "processor_unavailable";
    store.updatePaymentIntent(intent.id, { status: "failed" });
    store.createEvent(intent.merchantId, "payment_intent.failed", {
      id: intent.id,
      status: "failed",
      reason
    });
  }
};

const deliverWebhooks = async (eventId: string) => {
  const event = store.events.get(eventId);
  if (!event) return;
  const payload = {
    id: event.id,
    type: event.type,
    created_at: event.createdAt,
    data: event.data
  };
  const endpoints = store.listWebhookEndpoints(event.merchantId);
  for (const endpoint of endpoints) {
    try {
      await withRetries(
        () => deliverWebhook(endpoint.url, config.webhookSigningSecret, payload),
        config.maxRetryAttempts,
        config.baseBackoffMs,
        () => true
      );
    } catch {
      store.updateEventStatus(event.id, "failed");
      return;
    }
  }
  store.updateEventStatus(event.id, "delivered");
};

const runWorker = async () => {
  while (true) {
    const task = store.dequeueTask();
    if (!task) {
      await sleep(config.workerPollIntervalMs);
      continue;
    }
    if (task.type === "process_payment" && task.intentId) {
      await processPayment(task.intentId);
    }
    if (task.type === "deliver_webhooks" && task.eventId) {
      await deliverWebhooks(task.eventId);
    }
  }
};

runWorker();
